### PR TITLE
Move outbound call stats into OutRequest

### PR DESCRIPTION
**Pull request by Raynos**
This PR moves a lot of the stats code out of request.js and
directly into out_request.js instead.

It still preserves the semantics of only emitting stats
for a application driven request and not emitting stats
for retries.

This means we get stats for relay handler and for peer to peer
requests like ringpop.

r: @shannili @kriskowal

### CODE REVIEW

File: node/out_request.js [hunk 1]
```diff
@@ -35,9 +36,14 @@ function TChannelOutRequest(id, options) {
     self.responseEvent = self.defineEvent('response');
     self.finishEvent = self.defineEvent('finish');
 
+    assert(options.channel, 'channel required');
+
     self.logger = options.logger;
     self.random = options.random;
     self.timers = options.timers;
+    self.retryCount = options.retryCount;
```
Comment (ShanniLi) [inline node/out_request.js]: Why you need to pass in the retryCount? Shouldn't it be set to 0 when out request is initialized?

Comment (Raynos) [inline node/out_request.js]: we create one out request for every attempt. So this is either 0, 1, 2 etc. Every retry has a new OutRequest object.

Comment (ShanniLi) [inline node/out_request.js]: do we always retry on the same peer?

Comment (Raynos) [inline node/out_request.js]: never retry on same peer.

**Comment by ShanniLi** [comment]
lgtm

File: node/channel.js [hunk 2]
```diff
@@ -396,6 +396,7 @@ TChannel.prototype.request = function channelRequest(options) {
     if (opts.host || // retries are only between hosts
         opts.streamed // streaming retries not yet implemented
     ) {
+        opts.retryCount = 0;
```
Comment (kriskowal) [inline node/channel.js]: I suppose there’s no hope of this object ever being a hidden class.

**Comment by kriskowal** [comment]
:ok:

File: node/channel.js (hunk 2 above)
Comment (Raynos) [inline node/channel.js]: it can if we pre-allocate it and fix all the thingies.

**Pull request merged**